src/cargo/core/compiler/fingerprint.rs

```diff
@@ -923,57 +943,77 @@ impl Fingerprint {
                     },
                 ) => {
                     if *akey != *bkey {
-                        bail!("env vars changed: previously {}, now {}", bkey, akey);
+                        bail!(DirtyReason::EnvVarsChanged {
+                            old: bkey.clone(),
+                            new: akey.clone()
+                        })
                     }
                     if *avalue != *bvalue {
-                        bail!(
-                            "env var `{}` changed: previously {:?}, now {:?}",
-                            akey,
-                            bvalue,
-                            avalue
-                        )
+                        bail!(DirtyReason::EnvVarChanged {
+                            name: akey.clone(),
+                            old_value: bvalue.clone(),
+                            new_value: avalue.clone(),
+                        })
                     }
                 }
-                (a, b) => bail!(
-                    "local fingerprint type has changed ({} => {})",
-                    b.kind(),
-                    a.kind()
-                ),
+                (a, b) => bail!(DirtyReason::LocalFingerprintTypeChanged {
+                    old: b.kind(),
+                    new: a.kind()
+                }),
             }
         }
 
         if self.deps.len() != old.deps.len() {
-            bail!("number of dependencies has changed")
+            bail!(DirtyReason::NumberOfDependenciesChanged {
+                deps: DirtyReasonDeps {
+                    new: self.deps.clone(),
+                    old: old.deps.clone(),
+                }
+            })
         }
         for (a, b) in self.deps.iter().zip(old.deps.iter()) {
             if a.name != b.name {
-                let e = format_err!("`{}` != `{}`", a.name, b.name)
-                    .context("unit dependency name changed");
+                let e = format_err!(DirtyReason::UnitDependencyNameChanged {
+                    old: b.name.clone(),
+                    new: a.name.clone(),
+
+                    deps: DirtyReasonDeps {
+                        new: self.deps.clone(),
+                        old: old.deps.clone(),
+                    }
+                })
+                .context("unit dependency name changed");
```

Comment:
`compare_old_fingerprint` has a few other ways it can fail, such as not being able to read the fingerprint files, failing to deserialize the json, so the downcast is still needed, unless we introduce some other error type for that specifically.